java/flight/src/main/java/org/apache/arrow/flight/FlightClient.java

```diff
@@ -188,6 +188,11 @@ public ClientStreamListener startPut(FlightDescriptor descriptor, VectorSchemaRo
     ClientCallStreamObserver<ArrowMessage> observer = (ClientCallStreamObserver<ArrowMessage>)
         ClientCalls.asyncBidiStreamingCall(
             authInterceptor.interceptCall(doPutDescriptor, callOptions, channel), resultObserver);
+
+    // sync the schema from the vector structure,
+    // so that the server side can restore the vector schema root correctly
+    root.syncSchema();
```

Comment:
I think we should throw an exception here instead of trying to continue but i'm not sure of the flight model.  CC @lihalite

Reply:
Sure. Sounds reasonable. I have revised the code accordingly.

Reply:
Is there ever a case where you don't want the synchronized schema? IMO unless there's some use for it, we should always send the full and proper schema. Maybe @jacques-n could also chime in.

Reply:
@lihalite I wasn't sure of the code paths (which is why I CCed you), I thought this might occur after a schema was already sent?

Reply:
Ah sorry, I misunderstood. 

The schema isn't sent until the next line, in the call to `DictionaryUtils.generateSchemaMessages`: https://github.com/apache/arrow/blob/175ad654d45f4c014749b661ae3971ffe8f0512b/java/flight/src/main/java/org/apache/arrow/flight/DictionaryUtils.java#L49

Reply:
Thanks @liyafan82 sorry, no need to throw here this seems reasonable.

Reply:
@emkornfield Sure. Thanks.